docs/markdown/Using Pants/concepts/targets.md

```diff
@@ -388,3 +388,238 @@ shunit2_tests(
+
+Providing some example rule sets for the above example (see [rule actions](doc:targets#rule-actions) on how to mark a rule as "allow" or "deny"):
+
+```python
+# src/a/BUILD  (continued from previous example)
+__dependencies_rules__(
+  (
+    {"type": python_sources},  # We can use the target type unquoted when we don't need glob syntax
+    "src/a/**",  # May depend on anything from src/a/
+    "src/b/lib.py",  # May depend on specific file
+    "!*",  # May not depend on anything else. This is our "catch all" rule, ensuring there will never be any fall-through, which would've been an error
+  ),
+
+  # We need another rule set, in case we have non-python sources in here, to avoid fall-through.
+  # Sticking in a generic catch-all allow-all rule.
+  ({}, "*"),
+)
+
+
+# src/b/BUILD  (continued from previous example)
+__dependents_rules__(
+  (
+    (  # Using multiple selectors
+      {"type": "python_*", "tags":["any-python"]},
+      {"type": "*", "tags":["libs"]},
+      {"path": "special-cased.py"}
+    ),
+    (  # Grouping rules for readability
+      (  # Deny rules
+        "!tests/**",  # No tests
+        "!src/*/*/**", # Nothing deeply nested
+      ),
+      (  # Allow rules
+        "*",  # Allow everything else
+      ),
+    )
+  ),
+
+  # We need another rule set, in case we have non-python sources in here, to avoid fall-through.
+  # Sticking in a generic catch-all allow-all rule.
+  ({}, "*"),
+)
+```
+
+There are some syntactic sugar for selectors so they may be declared in a more concise text form rather than as a dictionary (this is also the form on which they are presented in messages from Pants, when possible). The syntax is `<type>(<tags>, ...)[<path>]`. With all parts optional, so `""` is a valid catch-all selector. Providing all the selectors from the previous example code block in string form for reference:
+```python
+python_sources  # {"type": python_sources}  -- target types works as strings when used bare
+"python_*(any-python)"  # {"type": "python_*", "tags":["any-python"]}
+"*(libs)"  # {"type": "*", "tags":["libs"]}
+"[special-cased.py]"  # {"path": "special-cased.py"}
+```
+
+Glob syntax
+-----------
+
+The visibility rules support a basic glob syntax, using `*` as a match anything non-recursively with the `**` variant being for a recursive match anything. All rules are matched until the end of the path it is being applied to, so if there is not trailing wildcard (`*` or `**`) the end of the rule glob will match the end of the path. This allows for matching on file types/names regardless of where in the project tree they are:
+```
+.py
+my_source.py
+my_*.py
+```
+
+Any leading wildcards may be used to emphasize this if desired, but will function the same, so the above is equvalent to (non-exhaustive list of alternatives):
+```
+*.py
+*/my_source.py
+*/my_*.py
```

Comment:
This is not currently accurate, but I think it should be!

That is, a rule glob that says `foo/bar` is translated to the regexp `foo/bar$`, and will match `no-goodfoo/bar`, where as I think the intention would've been to not match that, but `so-good/foo/bar` would be OK.

thoughts?

Reply:
Also, the `.py` case is currently an broken edge case, as it's treated as a invocation anchored glob, rather than a floating glob with a period in it. 

The fix: check for a lone period, or a one or more periods followed by a path sep only for the invocation anchor mode to kick in.

Reply:
Addressed by #17636